transcom/mymove
# B-21375 UB Allowance Calculation - INT

## [Agility ticket](https://www13.v1host.com/USTRANSCOM38/assetdetail.v1?number=B-21375)

Upstream Dependencies:
https://github.com/transcom/mymove/pull/13977

Downstream Dependencies:
https://github.com/transcom/mymove/pull/14102

## Summary

We are adding the UB allowance to the entitlements for an order. The UB allowance is calculated based off a combination of branch, grade, orders type, dependents authorized, whether or not it's an accompanied tour, and in the case of civilian pay grades - the number of dependents and whether they are under 12, or 12 and over.

Work in this PR:
- New table called ub_allowances
- Took data from the excel sheet attached to [B-21375](https://www13.v1host.com/USTRANSCOM38/assetdetail.v1?number=B-21375), pulled out the unique rows, and inserted them into ub_allowances to allow lookups based on branch, grade, orders type, dependents authorized, whether or not it's an accompanied tour
- Calculated the UB allowance in the cases of:
1. customer order creation
2. customer order amendment
3. SC creating orders on behalf of the customer
4. SC editing variables that would change the UB allowance (either in the Allowances section or the Orders section)
5. TOO editing variables that would change the UB allowance (either in the Allowances section or the Orders section)

The meat and potatoes of the logic is in pkg/models/entitlements.go, in the `GetUBWeightAllowance()` function.

### How to test

1. Set FEATURE_FLAG_UNACCOMPANIED_BAGGAGE to true.
2. Please download a copy of the [Uniformed Service Member Unaccompanied Baggage Lookup Table](https://www13.v1host.com/USTRANSCOM38/Default.aspx?Page=Widgets%2fDetails%2fAttachmentDetails&AssetWindow=true&AssetContext=Attachment%3a1022768)
3. Pick out a few random values and verify that they exist in the ub_allowances table and have the correct UB allowance value.
4. _Closely read the AC on the story_, especially the part about how to calculate civilian allowances. Please do the math for the civilian pay grade and the number of dependents and double check that your math matches mine.
5. Using the AC from Cameron's [B-21374](https://www13.v1host.com/USTRANSCOM38/assetdetail.v1?number=B-21374), go through every situation where the orders or allowances can be edited, and continue to recheck the orders.entitlement.ub_allowance value. Sometimes when it's edited, a new entitlement is added so you will have to retrace from the actual order down to the entitlement each time
6. Edit one field at a time - branch, order type, grade, dependents authorized, accompanied tour, etc. Try hard to break it - using combos from the spreadsheet, keep checking for accurate UB allowance values.
7. In the case of a combination that isn't valid to receive a UB entitlement, that's fine - just verify that the order.entitlement.ub_allowance is 0, as that's a valid value. You'll notice that the `GetUBWeightAllowance()` function doesn't return an error in most cases, as 0 is fine, it's just a situation where someone isn't entitled to a UB allowance.

Edge cases to pay attention to:
- Grade of E9 special senior enlisted should get the same UB allowance as E9s
- Space Force branch should get the same UB allowance as Air Force
- Any orders type other than Temporary Duty will be considered PCS, in regards to the spreadsheet values (per the AC on the feature)
- Local moves should get 0 for their UB allowance
- Max UB allowance for a family is 2000 lbs
